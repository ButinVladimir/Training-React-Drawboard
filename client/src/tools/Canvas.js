export const DEFAULT_DELTA_X = 0.0;
export const DEFAULT_DELTA_Y = 0.0;
export const DEFAULT_ZOOM = 1.0;
export const DEFAULT_ROTATION = 0.0;
export const DEFAULT_WIDTH = 500;
export const DEFAULT_HEIGHT = 500;
export const DEFAULT_CANVAS_COLOR = '#fff';
export const BORDER_COLOR = '#999';
export const BORDER_WIDTH = 3;

class Canvas {
  constructor() {
    this.deltaX = DEFAULT_DELTA_X;
    this.deltaY = DEFAULT_DELTA_Y;
    this.zoom = DEFAULT_ZOOM;
    this.rotation = DEFAULT_ROTATION;

    this.containerPositionX = 0;
    this.containerPositionY = 0;

    this.screenCanvasValue = null;
    this.screenContext = null;

    this.htmlCanvasElement = document.createElement('canvas');
    this.htmlCanvasElement.setAttribute('width', DEFAULT_WIDTH);
    this.htmlCanvasElement.setAttribute('height', DEFAULT_HEIGHT);

    this.htmlCanvasContext = this.htmlCanvasElement.getContext('2d');
    this.htmlCanvasContext.fillStyle = DEFAULT_CANVAS_COLOR;
    this.htmlCanvasContext.fillRect(0, 0, DEFAULT_WIDTH, DEFAULT_HEIGHT);

    this.bufferCanvasElement = document.createElement('canvas');
    this.bufferCanvasElement.setAttribute('width', DEFAULT_WIDTH);
    this.bufferCanvasElement.setAttribute('height', DEFAULT_HEIGHT);
    this.bufferCanvasContext = this.bufferCanvasElement.getContext('2d');
  }

  get screenCanvas() {
    return this.screenCanvasValue;
  }

  set screenCanvas(value) {
    this.screenCanvasValue = value;
    this.screenContext = value.getContext('2d');
  }

  outputCompletely() {
    this.screenContext.resetTransform();
    this.screenContext.clearRect(0, 0, this.screenCanvas.width, this.screenCanvas.height);

    this.show();
  }

  changeDelta(deltaX, deltaY) {
    this.hide();
    this.deltaX = deltaX;
    this.deltaY = deltaY;
    this.show();
  }

  changeZoom(zoom) {
    this.hide();
    this.zoom = zoom;
    this.show();
  }

  changeRotation(rotation) {
    this.hide();
    this.rotation = rotation;
    this.show();
  }

  applyTransformations() {
    this.screenContext.resetTransform();
    this.screenContext.translate(this.deltaX, this.deltaY);
    this.screenContext.rotate(this.rotation);
    this.screenContext.scale(this.zoom, this.zoom);
  }

  hide() {
    this.applyTransformations();

    this.screenContext.clearRect(
      -2 * BORDER_WIDTH, -2 * BORDER_WIDTH,
      DEFAULT_WIDTH + 4 * BORDER_WIDTH,
      DEFAULT_HEIGHT + 4 * BORDER_WIDTH,
    );
  }

  show() {
    this.applyTransformations();

    this.screenContext.strokeStyle = BORDER_COLOR;
    this.screenContext.lineWidth = BORDER_WIDTH;
    this.screenContext.strokeRect(0, 0, DEFAULT_WIDTH, DEFAULT_HEIGHT);

    this.screenContext.drawImage(this.htmlCanvasElement, 0, 0);
  }

  clearBuffer() {
    this.bufferCanvasContext.resetTransform();
    this.bufferCanvasContext.clearRect(
      0,
      0,
      this.bufferCanvasElement.width,
      this.bufferCanvasElement.height,
    );
  }

  updateAndShowBuffer(updateCb, ...args) {
    this.clearBuffer();

    this.bufferCanvasContext.scale(1 / this.zoom, 1 / this.zoom);
    this.bufferCanvasContext.rotate(-this.rotation);
    this.bufferCanvasContext.translate(-this.deltaX, -this.deltaY);
    this.bufferCanvasContext.translate(-this.containerPositionX, -this.containerPositionY);

    updateCb(this.bufferCanvasContext, ...args);

    this.show();
    this.screenContext.drawImage(this.bufferCanvasElement, 0, 0);
  }

  appendBuffer() {
    this.htmlCanvasContext.resetTransform();
    this.htmlCanvasContext.drawImage(this.bufferCanvasElement, 0, 0);
    this.clearBuffer();

    this.show();
  }
}

export default Canvas;
