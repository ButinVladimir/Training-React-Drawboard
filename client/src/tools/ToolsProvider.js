import BaseTool from '../components/tools/BaseTool';

class ToolsProvider {
  constructor() {
    this.registeredTools = new Map();
  }

  registerTool(tool) {
    if (tool === null || !(tool instanceof BaseTool)) {
      throw new TypeError('Tool must be an instance of any class extending BaseTool');
    }

    const { name } = tool;
    if (this.registeredTools.has(name)) {
      throw new Error(`${name} has been already registered within the provider`);
    }

    this.registeredTools.set(name, tool);
  }

  getTool(name) {
    if (name === null || !(typeof name === 'string')) {
      throw new TypeError('Tool name must be a string');
    }

    if (!this.registeredTools.has(name)) {
      throw new Error(`${name} is not registered within the provider`);
    }

    return this.registeredTools.get(name);
  }

  getToolsNames() {
    return Array.from(this.registeredTools.keys());
  }
}

export default ToolsProvider;
