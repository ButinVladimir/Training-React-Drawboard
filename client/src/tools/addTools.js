import MoveTool from './help/MoveTool';
import ZoomTool from './help/ZoomTool';
import RotateTool from './help/RotateTool';
import LineTool from './line/LineTool';
import RectangleTool from './rectangle/RectangleTool';
import CircleTool from './circle/CircleTool';

const addTools = (toolsProvider) => {
  toolsProvider.registerTool(MoveTool);
  toolsProvider.registerTool(ZoomTool);
  toolsProvider.registerTool(RotateTool);
  toolsProvider.registerTool(LineTool);
  toolsProvider.registerTool(RectangleTool);
  toolsProvider.registerTool(CircleTool);
};

export default addTools;
