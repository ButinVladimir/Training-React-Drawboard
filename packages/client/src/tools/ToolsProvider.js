import BaseTool from './BaseTool';

class ToolsProvider {
  constructor(canvas) {
    this.registeredTools = new Map();
    this.canvas = canvas;
  }

  registerTool(ToolConstructor) {
    if (ToolConstructor === null || !(ToolConstructor instanceof Function)) {
      throw new TypeError('ToolConstructor must be an constructor of any class extending BaseTool');
    }

    const { name } = ToolConstructor;
    if (name === null || !(typeof name === 'string')) {
      throw new TypeError('Tool name must be a string');
    }

    if (this.registeredTools.has(name)) {
      throw new Error(`${name} has been already registered within the provider`);
    }

    const tool = new ToolConstructor(this.canvas);

    if (tool === null || !(tool instanceof BaseTool)) {
      throw new TypeError('Tool must be an instance of any class extending BaseTool');
    }

    this.registeredTools.set(name, tool);
  }

  getTool(name) {
    if (name === null || !(typeof name === 'string')) {
      throw new TypeError('Tool name must be a string');
    }

    if (!this.registeredTools.has(name)) {
      throw new Error(`${name} is not registered within the provider`);
    }

    return this.registeredTools.get(name);
  }

  getToolsNames() {
    return Array.from(this.registeredTools.keys());
  }
}

export default ToolsProvider;
