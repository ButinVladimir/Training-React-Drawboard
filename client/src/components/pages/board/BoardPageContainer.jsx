import React, { Component } from 'react';
import PropTypes from 'prop-types';
import BoardPage from './BoardPage';
import ToolsProvider from '../../../tools/ToolsProvider';
import Canvas from '../../../tools/Canvas';
import ViewState from '../../../tools/ViewState';

class BoardPageContainer extends Component {
  constructor(props) {
    super(props);

    const { toolsProvider } = this.props;
    const selectedToolName = toolsProvider.getToolsNames()[0];

    this.onError = this.onError.bind(this);
    this.onDraw = this.onDraw.bind(this);
    this.onDrawSingleOperation = this.onDrawSingleOperation.bind(this);
    this.onGetImage = this.onGetImage.bind(this);
    this.onChangeTool = this.onChangeTool.bind(this);
    this.onResize = this.onResize.bind(this);
    this.onMouseDown = this.onMouseDown.bind(this);
    this.onMouseUp = this.onMouseUp.bind(this);
    this.onMouseMove = this.onMouseMove.bind(this);

    this.state = {
      errorMessage: '',
      selectedToolName,
    };
    this.selectedTool = toolsProvider.getTool(selectedToolName);

    this.canvasContainerRef = null;
    this.screenCanvasRef = null;
    this.bufferLoaded = false;
    this.buffer = [];
  }

  componentDidMount() {
    const {
      registerErrorHandlers,
      registerDrawHandler,
      registerGetImageHandler,
      emitGetImageEvent,
    } = this.props;

    registerErrorHandlers(this.onError);
    registerDrawHandler(this.onDraw);
    registerGetImageHandler(this.onGetImage);
    emitGetImageEvent();

    window.addEventListener('resize', this.onResize);
    this.onResize();
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.onResize);
  }

  onResize() {
    const { canvas } = this.props;
    const width = window.innerWidth - this.canvasContainerRef.offsetLeft;
    const height = window.innerHeight - this.canvasContainerRef.offsetTop;

    this.screenCanvasRef.setAttribute('width', width);
    this.screenCanvasRef.setAttribute('height', height);

    canvas.screenCanvas = this.screenCanvasRef;
    canvas.containerPositionX = this.canvasContainerRef.offsetLeft;
    canvas.containerPositionY = this.canvasContainerRef.offsetTop;
    canvas.outputCompletely();
  }

  onError(error) {
    this.setState({ errorMessage: error });
  }

  onDraw(viewStateObj, toolName, toolStateObj) {
    if (this.bufferLoaded) {
      this.onDrawSingleOperation(viewStateObj, toolName, toolStateObj);
    } else {
      this.buffer.push({ viewStateObj, toolName, toolStateObj });
    }
  }

  onGetImage(imageData) {
    const { canvas } = this.props;

    canvas.addPreloadedImage(imageData, () => {
      this.buffer.forEach(this.onDrawSingleOperation);
      this.bufferLoaded = true;
    });
  }

  onDrawSingleOperation(viewStateObj, toolName, toolStateObj) {
    const { canvas, toolsProvider } = this.props;
    const tool = toolsProvider.getTool(toolName);
    const viewState = ViewState.unserialize(viewStateObj);
    const toolState = tool.unserializeState(toolStateObj);

    canvas.applyState(tool.stateHandler, viewState, toolState);
  }

  onChangeTool(selectedToolName) {
    const { toolsProvider } = this.props;
    this.selectedTool = toolsProvider.getTool(selectedToolName);
    this.selectedTool.onSelect();

    this.setState({ selectedToolName });
  }

  onMouseDown(event) {
    this.selectedTool.screenHandler.onMouseDown(event);
  }

  onMouseUp(event) {
    const drawEventObj = this.selectedTool.screenHandler.onMouseUp(event);

    if (drawEventObj != null) {
      const { selectedToolName } = this.state;
      const { emitDrawEvent, canvas } = this.props;
      emitDrawEvent(
        canvas.viewState.serialize(),
        selectedToolName,
        drawEventObj,
      );
    }
  }

  onMouseMove(event) {
    this.selectedTool.screenHandler.onMouseMove(event);
  }

  render() {
    const { errorMessage, selectedToolName } = this.state;
    const { toolsProvider } = this.props;
    const toolsNames = toolsProvider.getToolsNames();

    const selectedToolOptions = this.selectedTool.optionsElement;

    return (
      <BoardPage
        errorMessage={errorMessage}
        toolsNames={toolsNames}
        selectedToolName={selectedToolName}
        selectedToolOptions={selectedToolOptions}
        onChangeTool={this.onChangeTool}
        onMouseDown={this.onMouseDown}
        onMouseUp={this.onMouseUp}
        onMouseMove={this.onMouseMove}
        setCanvasContainerRef={(ref) => { this.canvasContainerRef = ref; }}
        setScreenCanvasRef={(ref) => { this.screenCanvasRef = ref; }}
      />
    );
  }
}

BoardPageContainer.propTypes = {
  registerErrorHandlers: PropTypes.func.isRequired,
  registerDrawHandler: PropTypes.func.isRequired,
  registerGetImageHandler: PropTypes.func.isRequired,
  emitDrawEvent: PropTypes.func.isRequired,
  emitGetImageEvent: PropTypes.func.isRequired,
  toolsProvider: PropTypes.instanceOf(ToolsProvider).isRequired,
  canvas: PropTypes.instanceOf(Canvas).isRequired,
};

export default BoardPageContainer;
