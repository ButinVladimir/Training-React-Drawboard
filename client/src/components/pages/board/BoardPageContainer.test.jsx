import React from 'react';
import { mount } from 'enzyme';
import BoardPageContainer from './BoardPageContainer';
import ToolsProvider from '../../../tools/ToolsProvider';
import addTools from '../../../tools/addTools';
import Canvas from '../../../tools/Canvas';

describe('BoardPageContainer', () => {
  it('renders properly', () => {
    const canvas = new Canvas();
    const toolsProvider = new ToolsProvider(canvas);
    addTools(toolsProvider);
    const emitGetImageEventMock = jest.fn();
    const windowAddEventListenerSpy = jest.spyOn(window, 'addEventListener');
    const outputCompletelySpy = jest.spyOn(canvas, 'outputCompletely');

    const component = mount(<BoardPageContainer
      registerErrorHandlers={() => {}}
      registerDrawHandler={() => {}}
      registerGetImageHandler={() => {}}
      emitDrawEvent={() => {}}
      emitGetImageEvent={emitGetImageEventMock}
      toolsProvider={toolsProvider}
      canvas={canvas}
    />);
    const state = component.state();

    expect(component).toMatchSnapshot();
    expect(state).toMatchObject({
      errorMessage: '',
      selectedToolName: 'Move',
    });
    expect(emitGetImageEventMock).toHaveBeenCalled();
    expect(windowAddEventListenerSpy).toHaveBeenCalled();
    expect(outputCompletelySpy).toHaveBeenCalledTimes(1);
  });

  it('unnmounts properly', () => {
    const canvas = new Canvas();
    const toolsProvider = new ToolsProvider(canvas);
    addTools(toolsProvider);
    const windowRemoveEventListenerSpy = jest.spyOn(window, 'removeEventListener');

    const component = mount(<BoardPageContainer
      registerErrorHandlers={() => {}}
      registerDrawHandler={() => {}}
      registerGetImageHandler={() => {}}
      emitDrawEvent={() => {}}
      emitGetImageEvent={() => {}}
      toolsProvider={toolsProvider}
      canvas={canvas}
    />);

    component.unmount();

    expect(windowRemoveEventListenerSpy).toHaveBeenCalled();
  });


  it('handles onResize event', () => {
    const canvas = new Canvas();
    const toolsProvider = new ToolsProvider(canvas);
    addTools(toolsProvider);
    const outputCompletelySpy = jest.spyOn(canvas, 'outputCompletely');

    mount(<BoardPageContainer
      registerErrorHandlers={() => {}}
      registerDrawHandler={() => {}}
      registerGetImageHandler={() => {}}
      emitDrawEvent={() => {}}
      emitGetImageEvent={() => {}}
      toolsProvider={toolsProvider}
      canvas={canvas}
    />);

    window.dispatchEvent(new Event('resize'));
    expect(outputCompletelySpy).toHaveBeenCalledTimes(2);
  });

  it('handles onError event', () => {
    const errorMessage = 'An error occurred';
    const canvas = new Canvas();
    const toolsProvider = new ToolsProvider(canvas);
    addTools(toolsProvider);
    let onError;

    const component = mount(<BoardPageContainer
      registerErrorHandlers={(cb) => { onError = cb; }}
      registerDrawHandler={() => {}}
      registerGetImageHandler={() => {}}
      emitDrawEvent={() => {}}
      emitGetImageEvent={() => {}}
      toolsProvider={toolsProvider}
      canvas={canvas}
    />);

    onError(errorMessage);
    const state = component.state();

    expect(state).toMatchObject({ errorMessage });
  });

  it('handles onDraw event when buffer is not loaded', async () => {
    let onDraw;
    const canvas = new Canvas();
    const toolsProvider = new ToolsProvider(canvas);
    addTools(toolsProvider);

    const component = mount(<BoardPageContainer
      registerErrorHandlers={() => {}}
      registerDrawHandler={(cb) => { onDraw = cb; }}
      registerGetImageHandler={() => {}}
      emitDrawEvent={() => {}}
      emitGetImageEvent={() => {}}
      toolsProvider={toolsProvider}
      canvas={canvas}
    />);

    const instance = component.instance();

    expect(instance.buffer.length).toBe(0);

    const viewStateObj = {};
    const toolName = 'toolName';
    const toolStateObj = {};

    onDraw(viewStateObj, toolName, toolStateObj);

    expect(instance.buffer).toMatchObject([
      {
        viewStateObj,
        toolName,
        toolStateObj,
      },
    ]);
  });

  it('handles onDraw event when buffer is loaded', async () => {
    let onDraw;
    const canvas = new Canvas();
    const toolsProvider = new ToolsProvider(canvas);
    addTools(toolsProvider);

    const component = mount(<BoardPageContainer
      registerErrorHandlers={() => {}}
      registerDrawHandler={(cb) => { onDraw = cb; }}
      registerGetImageHandler={() => {}}
      emitDrawEvent={() => {}}
      emitGetImageEvent={() => {}}
      toolsProvider={toolsProvider}
      canvas={canvas}
    />);

    const instance = component.instance();
    instance.bufferLoaded = true;
    instance.onDrawSingleOperation = jest.fn();

    const viewStateObj = {};
    const toolName = 'toolName';
    const toolStateObj = {};

    onDraw(viewStateObj, toolName, toolStateObj);

    expect(instance.buffer.length).toBe(0);
    expect(instance.onDrawSingleOperation).toHaveBeenCalledWith(
      viewStateObj,
      toolName,
      toolStateObj,
    );
  });

  it('handles onGetImage event', async () => {
    let getImage;
    const canvas = new Canvas();
    const toolsProvider = new ToolsProvider(canvas);
    addTools(toolsProvider);

    const component = mount(<BoardPageContainer
      registerErrorHandlers={() => {}}
      registerDrawHandler={() => {}}
      registerGetImageHandler={(cb) => { getImage = cb; }}
      emitDrawEvent={() => {}}
      emitGetImageEvent={() => {}}
      toolsProvider={toolsProvider}
      canvas={canvas}
    />);

    canvas.addPreloadedImage = jest.fn();
    const instance = component.instance();
    instance.buffer = [1, 2, 3];
    instance.onDrawSingleOperation = jest.fn();

    await getImage();

    expect(canvas.addPreloadedImage).toHaveBeenCalled();
    expect(instance.onDrawSingleOperation).toHaveBeenCalledTimes(3);
    expect(instance.bufferLoaded).toBe(true);
  });
});
