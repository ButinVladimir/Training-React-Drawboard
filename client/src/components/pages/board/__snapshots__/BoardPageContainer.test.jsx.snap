// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardPageContainer renders properly 1`] = `
<div
  className="board-container"
  draggable={false}
>
  <div
    className="top-panel-container"
  >
    <div
      className="top-panel"
    >
      <div
        className="tool-select"
      >
        <select
          name="tool-name-options"
          onChange={[Function]}
          value="Move"
        >
          <option
            key="Move"
            value="Move"
          >
            Move
          </option>
          <option
            key="Zoom"
            value="Zoom"
          >
            Zoom
          </option>
          <option
            key="Rotate"
            value="Rotate"
          >
            Rotate
          </option>
          <option
            key="Brush"
            value="Brush"
          >
            Brush
          </option>
          <option
            key="Line"
            value="Line"
          >
            Line
          </option>
          <option
            key="Rectangle"
            value="Rectangle"
          >
            Rectangle
          </option>
          <option
            key="Circle"
            value="Circle"
          >
            Circle
          </option>
          <option
            key="Figure"
            value="Figure"
          >
            Figure
          </option>
        </select>
      </div>
      <div
        className="tool-options"
      >
        Array [
          <div
            className="tool-option-container"
          >
            <strong>
              Speed 
            </strong>
            <input
              name="speed"
              onChange={[Function]}
              type="text"
              value={1}
            />
            <span>
              (1)
            </span>
          </div>,
          <div
            className="tool-option-container"
          >
            <button
              name="set-default-speed"
              onClick={[Function]}
              type="button"
            >
              Set default speed
            </button>
          </div>,
          <div
            className="tool-option-container"
          >
            <button
              name="restore"
              onClick={[Function]}
              type="button"
            >
              Restore
            </button>
          </div>,
        ]
      </div>
    </div>
  </div>
  <div
    className="drawboard-canvas-container"
    id="drawboard-canvas-container"
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
  >
    <canvas
      draggable={false}
      id="drawboard-canvas"
    />
  </div>
</div>
`;
